### PR TITLE
fix(backend): add autopilot title fallback

### PR DESCRIPTION
**Pull request by Abhi1992002**
### Why / What / How

AutoPilot session titles currently depend on the aux AI title generation call. When the internal key is missing, the aux client is unavailable, or the AI response is empty, sessions can be left without a useful title and show up as an untitled chat.

This PR adds a deterministic backend fallback that derives a short title from the user's first message. The AI-generated title path is unchanged when the model call succeeds.

The fallback collapses whitespace, uses the first few words, caps titles at 50 characters, and falls back to `New chat` only when the opening message is empty after trimming. Empty AI responses still return the LLM response so title-generation cost recording can continue.

### Changes 🏗️

- Add `_fallback_title_from_message` for non-empty, human-readable session title fallback generation.
- Return fallback titles when aux title generation throws, the aux client is unavailable, choices are empty, message content is missing, or title content is blank.
- Update title-generation unit tests for success, missing-key/client failure, AI error, empty AI response, missing message content, and fallback formatting behavior.

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [ ] `poetry run prisma generate`
  - [ ] `poetry run pytest backend/copilot/service_unit_test.py -q -k "TestGenerateSessionTitle or TestUpdateTitleAsync or TestFallbackTitleFromMessage"` — 23 passed
  - [ ] `poetry run ruff check backend/copilot/service.py backend/copilot/service_unit_test.py`
  - [ ] `poetry run ruff format --check backend/copilot/service.py backend/copilot/service_unit_test.py`
  - [ ] `poetry run lint`

<details>
  <summary>Example test plan</summary>
  
  - [ ] Create from scratch and execute an agent with at least 3 blocks
  - [ ] Import an agent from file upload, and confirm it executes correctly
  - [ ] Upload agent to marketplace
  - [ ] Import an agent from marketplace and confirm it executes correctly
  - [ ] Edit an agent from monitor, and confirm it executes correctly
</details>

#### For configuration changes:

- [ ] `.env.default` is updated or already compatible with my changes
- [ ] `docker-compose.yml` is updated or already compatible with my changes
- [ ] I have included a list of my configuration changes in the PR description (under **Changes**)

<details>
  <summary>Examples of configuration changes</summary>

  - Changing ports
  - Adding new services that need to communicate with each other
  - Secrets or environment variable changes
  - New or infrastructure changes such as databases
</details>

### CODE REVIEW

**Comment by Abhi1992002** [comment]
🤖 Fixed in [ac56b3e2d](https://github.com/Significant-Gravitas/AutoGPT/commit/ac56b3e2d765197968652e599b99396188877774): aligned the mocked title-generation response with the new non-null fallback title contract and asserted the fallback title is persisted while cost recording still runs.

File: autogpt_platform/backend/backend/copilot/service.py [hunk 1]
```diff
@@ -676,9 +677,26 @@ async def _generate_session_title(
...
+        return "New chat"
+
+    words = cleaned.split()
+    title = " ".join(words[:6])
+    is_shortened = len(words) > 6
+    if len(title) > 50 or (is_shortened and len(title) > 47):
+        return title[:47] + "..."
+    if is_shortened:
+        return title + "..."
+    return title
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/copilot/service.py]: 🤖 Fixed in [502ef7505](https://github.com/Significant-Gravitas/AutoGPT/commit/502ef7505babce987a70eae82040dd320ff897d2): fallback title generation now strips server-injected display context before deriving the title, with regression coverage for memory and budget prefixes.

File: autogpt_platform/backend/backend/copilot/service.py [hunk 2]
```diff
@@ -787,9 +788,26 @@ async def _generate_session_title(
         title = title.strip().strip("\"'")
         if len(title) > 50:
             title = title[:47] + "..."
+    if not title:
+        title = _fallback_title_from_message(message)
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/service.py]: 🤖 🟡 **Nice to Have**: After this PR, `_generate_session_title` returns `tuple[str, ChatCompletion | None]` with a guaranteed non-empty title (fallback fires on every empty/error path at L776/L792). So in the caller `_update_title_async` at `service.py:968`, `if title and user_id:` has a dead `title` half — the guard is now misleading.

Scenario: a future reader sees `if title and user_id:` in the caller, assumes title-may-still-be-empty semantics, and adds an `else` branch that can never execute, or duplicates the fallback in the caller "just in case".

Question: drop `title and` in `_update_title_async` (and `if title:` at L790 here, which is also now guaranteed truthy after `_fallback_title_from_message`) so the post-PR contract is self-documenting?

File: autogpt_platform/backend/backend/copilot/service.py [hunk 3]
```diff
@@ -787,9 +788,26 @@ async def _generate_session_title(
...
     return title, response
 
 
+def _fallback_title_from_message(message: str) -> str:
+    cleaned = " ".join(strip_injected_context_for_display(message).split())
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/service.py]: 🟡 🤖 **Nice to Have**: `cleaned` is built only to be split a second time on the very next line.

Scenario:
1. User sends a 50KB message — `strip_injected_context_for_display(message).split()` materialises the full word list once.
2. `" ".join(...)` rebuilds a normalized string just so `cleaned.split()` on L800 can re-split it into the same word list.
3. Two O(n) passes + one extra allocation for nothing — we only ever consume `words[:6]`.

Question: collapse to `words = strip_injected_context_for_display(message).split(); if not words: return "New chat"` and drop the `cleaned` intermediate? Even better, use `split(maxsplit=6)` so a 50KB message doesn't materialise thousands of dead tokens.

File: autogpt_platform/backend/backend/copilot/service.py [hunk 4]
```diff
@@ -787,9 +788,26 @@ async def _generate_session_title(
...
+    if not cleaned:
+        return "New chat"
+
+    words = cleaned.split()
+    title = " ".join(words[:6])
+    is_shortened = len(words) > 6
+    if len(title) > 50 or (is_shortened and len(title) > 47):
+        return title[:47] + "..."
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/service.py]: 🤖 🟡 **Nice to Have**: Three independent magic numbers govern the truncation contract — `6` (word cap), `50` (char cap), `47` (char cap minus the 3-char ellipsis) — and the same `50`/`47` pair is also hard-coded at L789 in `_generate_session_title`.

Scenario:
1. Someone tightens the LLM-side cap from 50 → 40 at L789 without realising the fallback path here still uses 50/47.
2. AI-generated titles get cut at 40 chars but fallback titles stay at 50 — UI now shows inconsistent column widths, and the docstring claim "caps titles at 50 characters" quietly diverges between the two paths.

Question: lift `_TITLE_MAX_CHARS = 50`, `_TITLE_TRUNCATE_AT = 47`, `_TITLE_MAX_WORDS = 6` to module-level constants and reference them from both `_generate_session_title` and `_fallback_title_from_message` so the two paths can never drift?

File: autogpt_platform/backend/backend/copilot/service.py [hunk 5]
```diff
@@ -772,7 +773,7 @@ async def _generate_session_title(
         )
     except Exception as e:
         logger.warning(f"Failed to generate session title: {e}")
-        return None, None
+        return _fallback_title_from_message(message), None
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/service.py:784]: 🤖 🟡 **Nice to Have**: Exception path swallows the error and returns `(fallback_title, None)`, so the cost-recording branch in `_update_title_async` (which gates on `response is not None`) skips entirely.

Scenario:
1. OpenRouter accepts the request, starts streaming, and emits tokens (we get billed).
2. The connection drops mid-stream — `chat.completions.create(...)` raises a network exception caught here, we return `(fallback_title, None)`.
3. Caller sees `response is None`, skips `_record_title_generation_cost`, the spent credits never land on the user's ledger.

Question: pre-existing behaviour, but the updated docstring now claims "caller can still record the (paid-for) cost" specifically for the empty-LLM-response path — does the LLM-raises path warrant the same treatment (e.g., a zero-token marker record so partial-stream billing is at least observable), or is the title call cheap enough (~$0.0001) that silently dropping it is intentional?

File: autogpt_platform/backend/backend/copilot/service.py (hunk 2 above)
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/copilot/service.py]: 🤖 Fixed in [ee582e891](https://github.com/Significant-Gravitas/AutoGPT/commit/ee582e8910113f5e761784cf4a577fad6f1c8a93): removed the dead title guard half now that _generate_session_title always returns a non-empty title.

File: autogpt_platform/backend/backend/copilot/service.py (hunk 3 above)
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/copilot/service.py]: 🤖 Fixed in [ee582e891](https://github.com/Significant-Gravitas/AutoGPT/commit/ee582e8910113f5e761784cf4a577fad6f1c8a93): fallback title generation now splits the stripped display message once and reuses that word list for truncation.

File: autogpt_platform/backend/backend/copilot/service.py (hunk 4 above)
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/copilot/service.py]: 🤖 Fixed in [ee582e891](https://github.com/Significant-Gravitas/AutoGPT/commit/ee582e8910113f5e761784cf4a577fad6f1c8a93): extracted the title word limit, character limit, ellipsis, and truncated character limit into shared constants used by both generated and fallback title paths.

File: autogpt_platform/backend/backend/copilot/service.py (hunk 5 above)
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/copilot/service.py:784]: 🤖 Not applicable: this title call is non-streaming, so when chat.completions.create(...) raises there is no ChatCompletion object and no usage/cost payload for _record_title_generation_cost to persist. The existing response is not None gate records cost whenever the API returns a response, including empty-title responses.

**Comment by majdyz** [comment]
## 🧪 E2E Test Report

**PR:** #13141 — `fix(backend): add autopilot title fallback`
**Branch:** `abhimanyuyadav/secrt-2364-autopilot-session-title-fallback-when-openai-key-ai-is`
**Mode:** Native (poetry run app + pnpm dev, infra via docker compose deps)
**Worktree:** `/Users/majdyz/Code/AutoGPT6`

### Scope of change
Backend-only. Adds `_fallback_title_from_message()` and reworks `_generate_session_title` so it always returns a non-empty title (AI title on success, deterministic fallback derived from the user's first message on aux failure / empty AI response). `_update_title_async` drops the `if title` guard and now persists the title whenever `user_id` is set.

### Scenarios
| # | Scenario | Method | Expected | Actual | Result |
|---|----------|--------|----------|--------|--------|
| 1 | Targeted unit tests (`TestGenerateSessionTitle`, `TestUpdateTitleAsync`, `TestFallbackTitleFromMessage`) | `poetry run pytest -k …` | All pass | 24 passed | PASS |
| 2 | Happy path — first user message gets an AI-generated title persisted to `ChatSession.title` | Live `/api/chat/sessions` + `/stream`, assert via DB and `/api/chat/sessions?limit=…` | Title persisted to DB and surfaced in list | DB row `title="Marketing Competitor Research Agent"` from message `"Build me a marketing competitor research agent"`; visible in list endpoint | PASS |
| 3 | Aux-client exception → fallback used (mocked unit) | Mock `_get_aux_client` to raise, call `_generate_session_title` | Returns `(fallback_title, None)` derived from message | Returned `('Build me a competitor research dashboard...', None)` | PASS |
| 4 | AI returns empty title → fallback used (mocked unit) | Mock aux client returning `content=""` | Returns `(fallback_title, response)` so cost still records | Returned `('Plot revenue by quarter for last...', <response>)` | PASS |
| 5 | AI returns title → AI title used (no fallback, mocked unit) | Mock aux client returning `content="My Awesome Title"` | Returns `('My Awesome Title', response)` | Returned `('My Awesome Title', <response>)` | PASS |
| 6 | `_fallback_title_from_message` edge cases (whitespace collapse, injected-context strip, empty → "New chat", >6-word truncate, >50-char truncate, short pass-through) | Direct Python invocations | Match unit-test expectations | All 6 cases match | PASS |
| 7 | **E2E aux failure** — set `CHAT_AUX_API_KEY=<bad>` in .env, restart backend, send first message, verify fallback title persisted | Real HTTP, real backend, real bad aux key | Backend logs the 401 from aux client; DB row gets message-derived fallback title | Logs: `Failed to generate session title: Error code: 401 - Missing Authentication header`; DB row: `title="Plot revenue by quarter for last..."` (first 6 words + ellipsis) from message `"Plot revenue by quarter for last year for our internal sales team"` | PASS |

### Bugs found
None.

### Notes / observations
- `GET /api/chat/sessions/{id}` does NOT expose `title` in the response body; the title is surfaced through the list endpoint and persisted in the DB row. Pre-existing on `dev`, not a regression.
- Encountered pre-existing platform-cost-logging failure (`The column "trackingType" does not exist`) — unrelated to this PR. Title generation itself succeeded; only the cost-log side-effect errored.

### Verdict
**APPROVE.** Unit tests (24/24), happy-path live E2E, mocked aux-failure and empty-AI scenarios, and the headline gap — real bad aux key forcing a live 401 → message-derived fallback persisted — all green. The PR is doing exactly what it claims with no over-reach.

_Posted by majdyz local /pr-test._

**Review by majdyz — APPROVED** [review]
